loki/transform/transform_array_indexing.py

```diff
@@ -502,15 +502,20 @@ def normalize_array_shape_and_access(routine):
     """
     Shift all arrays to start counting at "1"
     """
-    def is_range_index(dim):
-        return isinstance(dim, sym.RangeIndex) and not dim.lower == 1
+    def is_explicit_range_index(dim):
+        # return False if assumed sized array or lower dimension equals to 1
+        return (isinstance(dim, sym.RangeIndex) and not dim.lower == 1 and not dim is None
+                    and not dim.lower is None and not dim.upper is None)
 
     vmap = {}
     for v in FindVariables(unique=False).visit(routine.body):
         if isinstance(v, sym.Array):
+            # skip if e.g., `array(len)`, passed as `call routine(array)`
+            if len(v.shape) != len(v.dimensions):
```

Comment:
Shouldn't the only case be that `v.dimensions is None`? Otherwise, `shape` and `dimensions` should match or there's something fishy anyway.